### PR TITLE
Feature vsan

### LINKED ISSUE #5966: Extend the vSphere plugin for vSAN.
### Proposal:

VMware vSAN is a software-defined storage solution which aggregates local or direct-attached data storage devices to create a single storage pool shared across all hosts in the vSAN cluster. vSAN is a distributed layer of software included in the ESXi hypervisor, and it is fully integrated with VMware vSphere.

I would like to extend the vSphere plugin for vSAN, so that vSAN users could be able to collect performance data, configuration metrics and some, similar to what vSphere plugin collects. This would be hugely helpful for whoever want to track, virtualize, or analyze vSAN metrics.

### Current behavior:

Nothing like this is currently written or available

### Desired behavior:

Should be able to collect vSAN data.
Metrics to collect should be configurable.

### Use case: 
By collecting and sending metrics to virtualizing tools, e.g., Wavefront, Grafana, Chronograf  (We are currently creating dashboards for theses platforms), it would be a big help for the following tasks: 
- Detect vSAN health and performance issue 
- Understand of current vSAN capacity, recync status, etc

### PR DESCRIPTION
**Pull request by YuxinZhou**
closes #5966

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### Changes made
- Made perf entities configurable for collection. Got default entities.
- Made vSAN data as new resource, so the configuration for vSAN stays consistent to other resources. 
- Collected vSAN health metrics
- Collected vSAN storage metrics 
- Collected vSAN resynchronization metrics 
- Added additional tags
- Added unit tests for the functions of cmmds and tags.

### Configuration 
Here is a sample configuration file.
```
[[inputs.vsphere]]
  interval = "600s"
  vcenters = [ "https://10.0.0.1/sdk" ]
  username = "username"
  password = "password"

  insecure_skip_verify = true
  force_discover_on_init = true

  # Exclude all other metrics
  vm_metric_exclude = ["*"]
  datastore_metric_exclude = ["*"]
  datacenter_metric_exclude = ["*"]
  host_metric_exclude = ["*"]
  cluster_metric_exclude = ["*"]
  # Enable vSan
  vsan_enabled = true
  # By default all supported entity will be included
  vsan_perf_metric_include = ["host-domclient", "cache-disk", "disk-group", "capacity-disk"]
  # vsan_perf_metric_exclude = ["*"]

  collect_concurrency = 5
  discover_concurrency = 5
```
By vsan_enabled = true, all vSAN metrics are collected by default. These include vSAN overall health, vSAN capacity and vSAN performance metrics for supported entity type. Since the supported entity types vary across releases, a list of supported entity will be fetched through GetSupportedEntityTypes API. To include or exclude performance entity type, vsan_perf_metric_include and vsan_perf_metric_exclude can be specified. However, notice that anything not in supported entity list will be filtered out.

The vSAN performance data is available every 5 minutes by default, so a recommend interval choice for the agent is between 300s to 3600s. The collector will query performance metrics for a duration from the time last sample collected to now. If the last sample time is unavailable or expires (timestamps expire in one hour), the default collecting duration is 15 minutes.

### Sample output

```
vsphere_cluster_vsan_performance_host-domclient,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,hostname=DC0_C0_H0,moid=domain-c7,source=Example-VSAN,vcenter=localhost:8898 latencyAvg=1307,congestion=0,oio=12,iopsWrite=19,latencyMaxUnmap=0,latencyStddev=1483,latencyAvgWrite=1709,writeCount=5722,clientCacheHitRate=0,writeCongestion=0,throughputUnmap=0,throughput=313682,ioCount=8298,latencyAvgRead=415,readCount=2576,clientCacheHits=0,numOio=107657,latencyMaxRead=10780,unmapCount=0,iopsRead=8,throughputRead=36561,throughputWrite=277121,readCongestion=0,unmapCongestion=0,iops=27,iopsUnmap=0,latencyAvgUnmap=0,latencyMaxWrite=13771 1557336300000000000
vsphere_cluster_vsan_performance_cluster-domcompmgr,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,uuid=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXX,vcenter=localhost:8898 latencyAvgRead=865,iopsRecWrite=0,oio=40,tputResyncRead=0,latAvgResyncRead=0,throughputRead=13116039,latencyAvgWrite=659,throughputRecWrite=0,congestion=0,latencyAvgRecWrite=0,iopsResyncRead=0,iopsRead=434,iopsWrite=444,throughputWrite=12907193 1557336300000000000
vsphere_cluster_vsan_performance_cluster-domclient,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,uuid=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXX,vcenter=localhost:8898 iopsWrite=137,throughputWrite=6415204,latencyAvgWrite=7689,congestion=0,oio=238,iopsRead=13,throughputRead=442438,latencyAvgRead=543 1557336300000000000
vsphere_cluster_vsan_summary,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,vcenter=localhost:8898 TotalBytesToSync=0i,TotalObjectsToSync=0i,TotalRecoveryETA=0i 1557336663000000000
vsphere_cluster_vsan_summary,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,vcenter=localhost:8898 OverallHealth=1i 1557336663000000000
vsphere_cluster_vsan_summary,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,vcenter=localhost:8898 TotalCapacityB=45129177395200i,FreeCapacityB=21377606853416i 1557336663000000000
```

### CODE REVIEW

**Comment by prydin** [comment]
@YuxinZhou looks like strings.Builder doesn't exist in Go 1.9. We're going to need an alternative.

**Comment by danielnelson** [comment]
We might drop Go 1.9 in the 1.12 release, but let confirm first.

**Comment by YuxinZhou** [comment]
@danielnelson Any conclusion on 1.12 version and when do you plan to release it? Thanks.

**Comment by prydin** [comment]
@danielnelson Did you come to a conclusion on the Go 1.9 support? We're getting errors in a library we're using (govmomi). Unfortunately, the version that has the functionality we need also uses string.Builder, which is not available in 1.9.

**Comment by danielnelson** [comment]
We are dropping Go 1.9 support in Telegraf 1.12 https://github.com/influxdata/telegraf/pull/6026

**Comment by prydin** [comment]
@YuxinZhou try merging @danielnelson 's latest changes. I think that should give you clean circle-ci tests.

**Comment by YuxinZhou** [comment]
@danielnelson Could you please take a look or assign reviewers? Once the pull request go public, we can plan further steps to publish the dashboard templates. Thanks.

**Comment by YuxinZhou** [comment]
@danielnelson @prydin Sorry for the repeated request, but is there any updates regarding this pull request? Also, it would be a great help if you can let me know the estimated time for reviewing and checking in.

**Comment by danielnelson** [comment]
@prydin Can you do an initial review?

**Comment by prydin** [comment]
> @prydin Can you do an initial review?

Will do over the next couple of days!

File: plugins/inputs/vsphere/vsphere.go
Comment (YuxinZhou) [inline plugins/inputs/vsphere/vsphere.go]: Will remove the flag

File: plugins/inputs/vsphere/vsan.go
Comment (YuxinZhou) [inline plugins/inputs/vsphere/vsan.go]: Do we need add limit here

File: plugins/inputs/vsphere/vsan.go
Comment (YuxinZhou) [inline plugins/inputs/vsphere/vsan.go]: should we put in endpoint.go?

Comment (YuxinZhou) [reply plugins/inputs/vsphere/vsan.go]: scale test?

File: plugins/inputs/vsphere/vsphere.go
Comment (YuxinZhou) [inline plugins/inputs/vsphere/vsphere.go]: The convention will be VSANPerfMetricExclude = ["*"]

**Comment by prydin** [comment]
@danielnelson @YuxinZhou I've reviewed the code and it looks good. I have two minor remarks:
1) In vsan.go there's one goroutine created per cluster. If you should encounter a vCenter with a very high number of clusters, there could be a situation where you're sending an excessive number of simultaneous requests to vCenter. I would recommend using the ThrottledExecutor that's part of the plugin and use the ```collect_concurrency``` setting for the maximum number of goroutines.
2) Drop the ```vsan_enabled``` and use ```vsan_perf_metric_exclude``` instead, since this is the convention in other plugins, as well as the core vSphere plugin.

Apart from this, I have no objections. Thank you @YuxinZhou for your contribution!

**Comment by YuxinZhou** [comment]
@prydin Thanks again for reviewing the code. I resolved all the comment issues.  @danielnelson Can you help review and check in the code? Thanks.

File: plugins/inputs/vsphere/vsan.go
Comment (danielnelson) [inline plugins/inputs/vsphere/vsan.go]: If these type assertions are wrong it will panic, I imagine they need to be checked.  The plugin should be able to handle bad or even malicious data without crashing.

File: plugins/inputs/vsphere/vsan.go
Comment (danielnelson) [inline plugins/inputs/vsphere/vsan.go]: Anywhere where you index into slice could panic if it is out of range, I don't see where this is checked.

File: plugins/inputs/vsphere/vsan.go
Comment (danielnelson) [inline plugins/inputs/vsphere/vsan.go]: To verbose, remove this line.  Also, on all log messages don't use `\t`.

File: plugins/inputs/vsphere/vsan.go
Comment (danielnelson) [inline plugins/inputs/vsphere/vsan.go]: On all error messages, use lowercase error message and space after `:`, the error string should say what happened:
```
acc.AddError(fmt.Errorf("error querying disk usage: %v", err))
```

File: plugins/inputs/vsphere/vsan.go
Comment (danielnelson) [inline plugins/inputs/vsphere/vsan.go]: If you need `cmmds` only on line 107, move this function down near there so that what you are doing is localized.  This will make it easier to follow the logic of why this is called.

File: plugins/inputs/vsphere/README-VSAN.md
Comment (YuxinZhou) [reply plugins/inputs/vsphere/README-VSAN.md]: Hi Daniel, thanks a lot for reviewing the code and leaving comments! For the readme file, the reasons we separate it are:
- the vSAN metrics don't follow the same structure as other vSphere resources, i.e, VM, cluster, memory. Putting them together might cause confusions.
- vSAN performance feature is only available after vSphere 5.5.
- Also, vSAN customers are mostly vSphere customers, but not necessarily vice versa. For the users who don't need to use the vSAN part, the document is already very long, and I prefer not to make it longer.
--
I will resolve the rest comments shortly. Thanks!
Best,
Yuxin

File: plugins/inputs/vsphere/vsan.go
Comment (YuxinZhou) [reply plugins/inputs/vsphere/vsan.go]: Otherwise, the certificate verification will fail. I will check with Govmomi guys to see if this is excepted.

File: plugins/inputs/vsphere/endpoint.go
Comment (YuxinZhou) [reply plugins/inputs/vsphere/endpoint.go]: The issue is vSAN has its own APIs and endpoints, it can't fit into collectResource func. We've also thought of creating a separate vSAN plugin, but we will need to copy code for connect, discovery, finder, etc.

File: plugins/inputs/vsphere/vsan.go
Comment (YuxinZhou) [reply plugins/inputs/vsphere/vsan.go]: Removed.

File: plugins/inputs/vsphere/vsan.go
Comment (YuxinZhou) [reply plugins/inputs/vsphere/vsan.go]: I would prefer keeping it for the troubleshooting purpose. If it is not the convention, maybe I can change it to "vSAN:"?

File: plugins/inputs/vsphere/vsan.go
Comment (YuxinZhou) [reply plugins/inputs/vsphere/vsan.go]: Resolved.

**Comment by YuxinZhou** [comment]
Hi Daniel,  
Most issues are resolved except two:  1. Merge the information in this README into the main vSphere readme. 2. Remove the [vSAN] part on all log messages.

Please see my reply under the above comments and let me know your opinion. Thank you for your time and really appreciate your help!

Best,
Yuxin

**Comment by YuxinZhou** [comment]
@danielnelson Do you have any plans of checking the code for the next release? We have been testing the code internally for a while and it looks quite stable up to now. Please let me know your plans. Thanks

**Comment by danielnelson** [comment]
Yes, we are planning to take a closer look for the next release.

**Comment by YuxinZhou** [comment]
> Yes, we are planning to take a closer look for the next release.

Thanks a lot!

**Comment by ambekarnikhil** [comment]
@danielnelson Thanks for the review. I saw a new release(1.12.3) recently. Please let us know your plans for the merge.

**Comment by glinton** [comment]
@ambekarnikhil we should have clarified, "next release" means next [minor release](https://semver.org/) this isn't on the github milestone for 1.13, but it sounds like it is penciled in for that release.

**Comment by puckpuck** [comment]
@glinton is the plan still to include this with 1.13?

**Comment by puckpuck** [comment]
@glinton and @danielnelson what do we need to do to get this PR merged?  Seems like everything is ready, but it's not part of any upcoming release.

**Comment by prydin** [comment]
@YuxinZhou @danielnelson @glinton @puckpuck  Can I help here? It looks like this has stalled a bit and we're still seeing a steady stream of requests for this functionality.

**Comment by YuxinZhou** [comment]
@prydin Thanks a lot for offering help. @ambekarnikhil is taking over the task now. And I guess he already synced up with you?

**Comment by Spookman** [comment]
Hi, is this feature still actively being worked on, vSAN is now and going to be even more so core to VMWare, especially with kubernetes integrations and VCF, from a VM perspective it's only the disk metrics that are missing when you move a VM from a SAN storage to a vSAN storage, even if we only got those stats reported through that would be a help.

**Comment by ambekarnikhil** [comment]
Hi @Spookman, We are waiting for review and getting merge approved from @danielnelson. 

@danielnelson Can you review the merge request and let us know if we can merge it? Thanks!

**Comment by Spookman** [comment]
Is there any movement on this please ?

**Comment by Spookman** [comment]
Hello everyone, is there any traction on this ?, this plugin is going to be obsolete soon if not kept up to date, vsphere now has contain storage and kubernetes we are going to see environments transformed and this plugin will be useless if it can only collect half the metrics.

**Comment by eschek87** [comment]
I hope that the vSAN feature will be releases / integrated soon!!!

**Comment by ambekarnikhil** [comment]
Hi, Sorry for the delay. I am waiting for the review from telegraf team to help us merge the codes.

**Comment by jorgedlcruz** [comment]
Hi guys, 
I would love to have this as well integrated, seems that all the code is there. Adding vSAN support is key for VMC, VCF, and so many others, vSAN is the de facto storage for VMware, please can we take a look into this?

**Comment by sjwang90** [comment]
@ambekarnikhil It doesn't look like @danielnelson's comments on this PR were ever addressed. Can you address his comments and we'll get more reviews done from the Telegraf team and @prydin.

Thanks!

**Comment by rolfbartels** [comment]
Hi, Is there anyone working on this, this is quite an old issue and vSAN has now been around a while, would be great to have this included.

**Comment by ambekarnikhil** [comment]
Most of the comments are resolved. @prydin @YuxinZhou Can you help mark the comments as resolved.

For the sibling plugin, since most of the codes will be duplicated and this is part of vsphere we will not create a new plugin

File: plugins/inputs/vsphere/vsan.go
Comment (srebhan) [inline plugins/inputs/vsphere/vsan.go:205]:
```suggestion

	var clusterCmmds Cmmds
```

File: plugins/inputs/vsphere/vsan.go
Comment (srebhan) [inline plugins/inputs/vsphere/vsan.go:214]:
```suggestion
		cmmdsMap[entity.UUID] = entity
```

File: plugins/inputs/vsphere/vsan.go
Comment (srebhan) [inline plugins/inputs/vsphere/vsan.go:228]: Why do you need this copy. Can't you use `start` directly?

File: plugins/inputs/vsphere/vsan.go
Comment (srebhan) [inline plugins/inputs/vsphere/vsan.go:242]:
```suggestion
		
		perfSpecs := []vsantypes.VsanPerfQuerySpec {
			vsantypes.VsanPerfQuerySpec{
				EntityRefId: fmt.Sprintf("%s:*", entityRefId),
				StartTime:   &start,
				EndTime:     &end,
			},
		}
```

File: plugins/inputs/vsphere/vsan.go
Comment (srebhan) [inline plugins/inputs/vsphere/vsan.go:377]: First of all please check the potential error. Second, please just remove the call to this function instead of logging this over and over. Third, do not use `log`.

File: plugins/inputs/vsphere/vsan.go
Comment (srebhan) [inline plugins/inputs/vsphere/vsan.go:393]: This will panic on values not containing the expected pattern.

**Review by srebhan — CHANGES_REQUESTED** [review]
Hey @YuxinZhou and @ambekarnikhil thanks for keep this going. I tried to mark all comments from Daniel resolved and kept those in my review to tack the current state.

The code looks quite complex and I have some comments in the code. Please try to first move the code to Go modules and rebase it on a current master. After that, please try to get rid of as many global variables as possible and skip collection of endpoints with insufficient versions as early as possible...

Looking forward to the next round. Keep up the good work!

**Comment by powersj** [comment]
Hi,

This PR has not been touched in a while and as a result I am going to close it. We are happy to review again once someone picks this back up with the suggested changes and rebased on master.

Thanks!

**Pull request closed without merging**
